tools/filter.py

```diff
@@ -0,0 +1,150 @@
+# This file is part of the EESSI build-and-deploy bot,
+# see https://github.com/EESSI/eessi-bot-software-layer
+#
+# The bot helps with requests to add software installations to the
+# EESSI software layer, see https://github.com/EESSI/software-layer
+#
+# author: Thomas Roeblitz (@trz42)
+#
+# license: GPLv2
+#
+import re
+
+from collections import namedtuple
+from pyghee.utils import log
+
+
+FILTER_COMPONENTS = ['architecture', 'instance', 'job', 'repository']
+
+Filter = namedtuple('Filter', ('component', 'pattern'))
+
+
+class EESSIBotActionFilterError(Exception):
+    pass
+
+
+class EESSIBotActionFilter:
+    def __init__(self, filter_string):
+        """
+        EESSIBotActionFilter constructor
+
+        Args:
+            filter_string (string): string containing whitespace separated filters
+        """
+        self.action_filters = []
+        for _filter in filter_string.split():
+            try:
+                self.add_filter_from_string(_filter)
+            except EESSIBotActionFilterError:
+                raise
+            except Exception as err:
+                log(f"Unexpected err={err}, type(err)={type(err)}")
+                raise
+
+    def clear_all(self):
+        self.action_filters = []
+
+    def add_filter(self, component, pattern):
+        """
+        Adds a filter
+
+        Args:
+            component (string): any prefix of 'architecture', 'instance', 'job' or 'repository'
+            pattern (string): regex that is applied to a string representing the component
+        """
+        # check if component is supported
+        full_component = None
+        for cis in FILTER_COMPONENTS:
+            if cis.startswith(component):
```

Comment:
Right now we don't have filter components that partially overlap, but what if we introduce one?

At the very least we should add a comment above definition of `FILTER_COMPONENTS` to warn about this, and maybe this code needs to be made a bit more intelligent too (collect all prefix matches, only conclude we found match if there's only one match)

Reply:
True. I can't think of two components with the same prefix ... Jokes aside, I added a note at the definition of `FILTER_COMPONENTS` and also to `add_filter` (that it assumes that no two filter share the same prefix).

Reply:
OK, good enough for now